### ISSUE TITLE
Support x86 in new instruction selection framework

### ISSUE THREAD

**Issue by tschneidereit** [bytecodealliance/wasmtime#1980]
We don't currently have fully working x86 support, but we have a backend in-tree that's largely completely. However, it uses the old instructions selection framework, and as described in #1936, Cranelift is in the process of switching to a new one. While the old framework won't be removed until quite a bit later, that means that not much work will go into it anymore, and that any investments into backends based on it will become obsolete.

We should investigate the best approach to supporting x86 in the new framework.

There's active work going on to implement x64 support in the new framework, but no current effort around x86. In #1789 @whitequark mentions that she'd be happy to contribute to x86 support in the new framework, but only if the general support for targeting x86 exists.

@bnjbvr @julian-seward1, can you comment on what the best approach to take here would be? I understand that this isn't something that's high on your priority list, so as a first step we should just get a clearer picture of what this could look like, and the effort required.

**Comment by bnjbvr** [comment]
Thanks for opening an issue!

## Sharing code

The current encoding machinery is sufficient to support all the x86 32 and 64 bits instructions up to SSE4.2 (i.e., nothing has been done for VEX yet). Most encodings can be reused on 32 bits (of course, without the 64-bit mode); some specific encodings and addressing modes are not available in 32-bits mode (e.g. RIP-relative addressing).

Ideally, most of the encoding code could be shared between the two architectures, with an `Inst` enum for instructions that are common to both architectures; each of x86_64 and x86-32bits would then have their own `Inst` enum as well, with a particular `Shared` variant that would be dynamically dispatch most queries / calls to the shared `Inst` enum. It would mean that the 64-bits and 32-bits backends would have special code to handle the shared encodings (e.g. when printing out the instructions and registers, when handling supported operand sizes,...) so it might become a bit messy. Moreover, the code for these two architectures would be scattered in more places. Spidermonkey has tried to enforce code sharing between these two architectures, and it's made the code difficult to follow sometimes (it tends to force low-level abstractions that add unnecessary layers to the code). But we wouldn't need to duplicate work.

Alternatively, we could just duplicate all the Inst variants, so that each target has precisely what it can do. Even if this implies some code duplication, this should stay reasonably well-contained over time: code emitting encodings tends to remain stable over time, once it's been written (and maybe fixed once or twice).

Since this encoding code is still in the process of being written, it might be good to wait for stabilization before jumping into implementing a new target derived from it.

## Legalizing IR

At this point, there's not a good story for transforming sequences of operations involving types that are not natively supported by the platform: be it i128 or aarch64/x64, or i64 on on x86 32 bits. We'd probably need something like lightweight legalizations to make this maintainable over time.

 ## ABIs

32-bits ABIs would be one more work item to do, since the ABIs are quite different. This isn't quite complicated, but it requires careful testing to be sure not to make an error there.

In a nutshell, there's nothing that's too complicated. I'm not quite sure of the best way to share code yet.

**Comment by guest271314** [comment]
> Error: Sorry! Wasmtime currently only provides pre-built binaries for x86_64 architectures.

**Comment by guest271314** [comment]
It would be helpful and informative if the README.md at https://github.com/bytecodealliance/wasmtime including an unambiguous and conspicuous header and language that states x86 32 bit architecture is not (currently) supported, so that users do not run

`$ curl https://wasmtime.dev/install.sh -sSf | bash`

or try https://wasi.dev/polyfill/ using instructions at https://github.com/bytecodealliance/wasmtime/blob/main/docs/WASI-tutorial.md expecting the code to work.

**Comment by bjorn3** [comment]
I think this can be closed.

**Comment by whitequark** [comment]
How can I test 32-bit x86 support? `--target i686-unknown-linux-gnu` doesn't seem to work for example:
```
thread 'main' panicked at 'host machine is not a supported target: "unsupported architecture"', crates/jit/src/native.rs:6:33
```

**Comment by bjorn3** [comment]
The new x64 backend doesn't have 32bit support yet.

**Comment by whitequark** [comment]
That's exactly what this issue is about, no?

**Comment by bjorn3** [comment]
Oops, interpreted x86 as x86_64 instead of 32bit x86.

**Comment by jain98** [comment]
@tschneidereit @whitequark @cfallin @bjorn3 I would be interested in picking this up. I still need to dive deeper to come up with a concrete plan, but would one of you (or someone you could connect me to) be willing to mentor me?

**Comment by jameysharp** [comment]
Hi @jain98! We're always happy to answer questions and help however we can. I'd encourage you to join the `#cranelift` stream at https://bytecodealliance.zulipchat.com/ as a good place to ask questions.

Since all our current backends are 64-bit, we haven't necessarily designed everything to support 32-bit targets well. We'd like to fix that, but in the meantime just be aware that you may find some things are harder than they should be. There's been recent discussion about that in #5572, and the advice there applies here too.

**Comment by jain98** [comment]
Hi @jameysharp, thanks for replying and a link to a recent discussion. I've already joined the cranelift zulip chat, but figured I'd start here as it seemed more intentional. 

It at least looks like there's already an existing backend for x86-32, no? If yes, could you could point me to it? That might be a good starting point for me.

**Comment by jameysharp** [comment]
There used to be an x86-32 backend, but it was unmaintained and eventually got in the way of working on the rest of Cranelift, so it was deleted. If I'm reading the history correctly there have been at least two complete rewrites of how Cranelift backends are written since the last time that backend worked, so digging it up out of the git history probably won't help you much. Instead I'd suggest looking at the current x86-64 backend, which is mostly in `cranelift/codegen/src/isa/x64/`.

**Comment by alexcrichton** [comment]
For those following this issue Wasmtime now supports 32-bit x86 via the [Pulley interpreter](https://github.com/bytecodealliance/rfcs/blob/main/accepted/pulley.md) which should now be able to run most wasm programs. Pulley is not 100% finished yet in that we still have some polish to do, but it should be suitable for testing at this time. 

General support for Wasmtime on 32-bit x86 architectures should be "solved" now and if there are any issues please feel free to open a new issue, and this issue specifically will continue to track the addition of a 32-bit x86 backend for Cranelift.

**Comment by whitequark** [comment]
Oh, fantastic! From my perspective Pulley is probably good enough, since the failure case I'm trying to avoid is "YoWASP does not work"; if it works but is slow, I can print a message recommending to upgrade to a 64-bit Python.

(Is there a way to detect that Pulley is being used?)

**Comment by alexcrichton** [comment]
Not currently, but it'd be reasonable to add one for sure

**Comment by whitequark** [comment]
I'd really like to have that, since as I understand there is a ~10x perf degradation, which is fine in that it enables the capability, but this being someone's only impression without understanding that it's unusually slow wouldn't be good.

**Comment by RyanHope** [comment]
So, 2026 still no 32bit support?

**Comment by RyanHope** [comment]
Is this going anywhere?
